File: java/revenj-servlet/src/main/java/org/revenj/server/commands/reporting/PopulateReport.java
package org.revenj.server.commands.reporting;

import org.revenj.patterns.*;
import org.revenj.security.PermissionManager;
import org.revenj.serialization.Serialization;
import org.revenj.server.CommandResult;
import org.revenj.server.ReadOnlyServerCommand;

import java.io.IOException;
import java.security.Principal;
import java.util.Optional;

public class PopulateReport implements ReadOnlyServerCommand {

	private final DomainModel domainModel;
	private final PermissionManager permissions;

	public PopulateReport(
			DomainModel domainModel,
			PermissionManager permissions) {
		this.domainModel = domainModel;
		this.permissions = permissions;
	}

	public static final class Argument<TFormat> {
		public TFormat Data;
		public String ReportName;

		public Argument(TFormat data, String reportName) {
			this.Data = data;
			this.ReportName = reportName;
		}

		@SuppressWarnings("unused")
		private Argument() {
		}
	}

	@SuppressWarnings("unchecked")
	@Override
	public <TInput, TOutput> CommandResult<TOutput> execute(
			ServiceLocator locator,
			Serialization<TInput> input,
			Serialization<TOutput> output,
			TInput data,
			Principal principal) {
		Argument<TInput> arg;
		try {
			arg = input.deserialize(data, Argument.class, data.getClass());
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
		Optional<Class<?>> manifest = domainModel.find(arg.ReportName);
		if (!manifest.isPresent()) {
			return CommandResult.badRequest("Couldn't find report type: " + arg.ReportName);
		}
		if (!permissions.canAccess(manifest.get(), principal)) {
			return CommandResult.forbidden(arg.ReportName);
		}
		if (!Report.class.isAssignableFrom(manifest.get())) {
			return CommandResult.badRequest("Specified type is not a report: " + arg.ReportName);
		}
		final Report report;
		try {
			report = (Report)input.deserialize(arg.Data, manifest.get());
		} catch (IOException e) {
			return CommandResult.badRequest("Error deserializing report: " + arg.ReportName + ". Reason: " + e.getMessage());
		}
		final Object result = report.populate(locator);
		try {
			return CommandResult.success("Report populated", output.serialize(result));
		} catch (IOException e) {
			return new CommandResult<>(null, "Error serializing result.", 500);
		}
	}
}

File: scala/revenj-akka/src/main/scala/net/revenj/server/commands/reporting/PopulateReport.scala
package net.revenj.server.commands.reporting

import net.revenj.patterns._
import net.revenj.serialization.Serialization
import net.revenj.server.commands.reporting.PopulateReport.Argument
import net.revenj.server.{CommandResult, ReadOnlyServerCommand}

import scala.concurrent.Future

class PopulateReport(domainModel: DomainModel) extends ReadOnlyServerCommand {

  override def execute[TInput, TOutput](
    locator: ServiceLocator,
    input: Serialization[TInput],
    output: Serialization[TOutput],
    data: TInput): Future[CommandResult[TOutput]] = {

    val arg = input.deserializeRuntime(data, classOf[Argument[TInput]], data.getClass)
    lazy val manifest = domainModel.find(arg.get.ReportName)
    if (!arg.isSuccess) {
      CommandResult.badRequest(arg.failed)
    } else if (manifest.isEmpty) {
      CommandResult.badRequest(s"Unable to find report type: ${arg.get.ReportName}")
    } else if (!classOf[Report[_]].isAssignableFrom(manifest.get)) {
      CommandResult.badRequest(s"Specified type is not a report: ${arg.get.ReportName}")
    } else {
      val report = input.deserializeRuntime[Report[_]](arg.get.Data, manifest.get)
      if (!report.isSuccess) {
        CommandResult.badRequest(s"Error deserializing report: ${arg.get.ReportName}. Reason: ${report.failed.get.getMessage}")
      } else {
        import scala.concurrent.ExecutionContext.Implicits.global
        report.get.populate(locator).map { result =>
          val response = output.serializeRuntime(result)
          if (response.isSuccess) {
            CommandResult[TOutput](Some(response.get), result.toString, 200)
          } else {
            CommandResult[TOutput](None, "Error serializing result.", 500)
          }
        }
      }
    }
  }
}

object PopulateReport {

  case class Argument[TFormat](
    ReportName: String,
    Data: TFormat)

}